# Ejercicio 2

El ejercicio en una cadena de Markov, porque describe una secuencia de eventos _(estados)_ donde la probabilidad del futuro depende exclusivamente del estado actual.

La cadena de Markov puede ser visto como un grafo dirigido y ponderado:
**Nodos:** Estados del 1 al 7
**Aristas:** Transiciones posibles entre los estados (anterior-mismo-siguiente)
**Ponderación:** Probabilidades ($0 \le P_{ij} \le 1$)

La matriz de transición (P) es una matriz estocástica:

| Estado Actual | Estado Anterior | Mismo Estado | Estado Siguiente |
| :---: | :---: | :---: | :---: |
| 1 | 0 | 0.95 | 0.05 |
| 2 | 0.40 | 0.30 | 0.30 |
| 3 | 0.45 | 0.10 | 0.45 |
| 4 | 0.40 | 0.20 | 0.40 |
| 5 | 0.45 | 0.10 | 0.45 |
| 6 | 0.30 | 0.30 | 0.40 |
| 7 | 0.05 | 0.95 | 0 |


In [2]:
#Librerías necesarias:

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from matplotlib.gridspec import GridSpec

In [ ]:
#Matriz de transición

P = np.array([
    # E1    E2    E3    E4    E5    E6    E7   ← Estado destino
    [0.95, 0.05, 0.00, 0.00, 0.00, 0.00, 0.00],  # Desde Estado 1
    [0.40, 0.30, 0.30, 0.00, 0.00, 0.00, 0.00],  # Desde Estado 2
    [0.00, 0.45, 0.10, 0.45, 0.00, 0.00, 0.00],  # Desde Estado 3
    [0.00, 0.00, 0.40, 0.20, 0.40, 0.00, 0.00],  # Desde Estado 4 (Estado inicial)
    [0.00, 0.00, 0.00, 0.45, 0.10, 0.45, 0.00],  # Desde Estado 5
    [0.00, 0.00, 0.00, 0.00, 0.30, 0.30, 0.40],  # Desde Estado 6
    [0.00, 0.00, 0.00, 0.00, 0.00, 0.05, 0.95],  # Desde Estado 7
])

#Como es Markov cada fila debe sumar 1, lo verificamos:
print("PASO 1: VERIFICACIÓN DE LA MATRIZ DE TRANSICIÓN P")
print("\nSuma de cada fila (todas deben ser exactamente 1.0):")
for i, suma in enumerate(P.sum(axis=1)):
    print(f"  Fila del Estado {i+1}: {suma:.2f} ✓")

PASO 1: VERIFICACIÓN DE LA MATRIZ DE TRANSICIÓN P

Suma de cada fila (todas deben ser exactamente 1.0):
  Fila del Estado 1: 1.00 ✓
  Fila del Estado 2: 1.00 ✓
  Fila del Estado 3: 1.00 ✓
  Fila del Estado 4: 1.00 ✓
  Fila del Estado 5: 1.00 ✓
  Fila del Estado 6: 1.00 ✓
  Fila del Estado 7: 1.00 ✓


In [4]:
# Nombres de los estados (para etiquetas en gráficos)
nombres_estados = [
    "1: Muy\nInsatisfecho",
    "2: Insatisfecho",
    "3: Lig.\nInsatisfecho",
    "4: Neutral",
    "5: Lig.\nSatisfecho",
    "6: Satisfecho",
    "7: Muy\nSatisfecho"
]


## 2.1. Simulación de caminata aleatoria

Para la simulación se hace uso de una función para ejecutar la simulación, y la simulación la repetimos 10 veces haciendo uso del método Monte Carlo para luego observar tendencias estadísticas.

In [ ]:
#Función de simulación de la cadena de Markov (camianta aleatoria)
def simular_caminata(estado_inicial, num_pasos, matriz_P): #(int (1-7), int, np.array)
   
    # Convertimos a índice Python (0-6)
    estado_actual = estado_inicial - 1

    # Guardamos el recorrido completo
    recorrido = [estado_actual + 1]  # el +1 devuelve a escala 1-7

    for _ in range(num_pasos):
        # Tomamos las probabilidades de transición desde el estado actual
        # Matemáticamente: esta es la fila P[estado_actual, :]
        probabilidades = matriz_P[estado_actual, :]

        # Elegimos el siguiente estado según esas probabilidades
        # np.random.choice(7) elige un número entre 0 y 6,
        # con las probabilidades dadas
        estado_actual = np.random.choice(7, p=probabilidades)

        # Guardamos en escala 1-7
        recorrido.append(estado_actual + 1)

    return recorrido

In [14]:
#Simulación de la caminata usando la función
estado_inicial = 4    # Neutral
num_pasos      = 100  # Pasos por simulación
num_simulaciones      = 10   # Número de repeticiones


print("TAREA 1: SIMULACIÓN DE 10 CAMINATAS (100 pasos c/u)")
print(f"Estado inicial: {estado_inicial} (Neutral)")

simulaciones   = []
estados_finales = []

for i in range(num_simulaciones):
    recorrido = simular_caminata(estado_inicial, num_pasos, P)
    simulaciones.append(recorrido)
    estados_finales.append(recorrido[-1])
    print(f"  Simulación {i+1:2d}: estado final = {recorrido[-1]}")

print(f"\nResumen de estados finales: {estados_finales}")
print(f"Promedio del estado final:  {np.mean(estados_finales):.2f}")
print(f"Estado final más frecuente: {max(set(estados_finales), key=estados_finales.count)}")

TAREA 1: SIMULACIÓN DE 10 CAMINATAS (100 pasos c/u)
Estado inicial: 4 (Neutral)
  Simulación  1: estado final = 7
  Simulación  2: estado final = 7
  Simulación  3: estado final = 1
  Simulación  4: estado final = 4
  Simulación  5: estado final = 1
  Simulación  6: estado final = 4
  Simulación  7: estado final = 7
  Simulación  8: estado final = 7
  Simulación  9: estado final = 7
  Simulación 10: estado final = 1

Resumen de estados finales: [7, 7, 1, 4, 1, 4, 7, 7, 7, 1]
Promedio del estado final:  4.60
Estado final más frecuente: 7


## 2.2. Tendencia del cliente

El objetivo es responde  ¿A donde tiende a terminar el cliente con un nivel de satisfacción neutral después de un tiempo?

Para tener una mejor estimación estadística se aumenta el número de simulaciones, usando el método Monte Carlo.

In [ ]:
num_simulaciones_grande = 500

print("TAREA 2: TENDENCIA DE CAMBIO DE LOS ESTADOS (500 simulaciones)")
print("=" * 60)

estados_finales_grande = []

for _ in range(num_simulaciones_grande):
    recorrido = simular_caminata(estado_inicial, num_pasos, P)
    estados_finales_grande.append(recorrido[-1])

# Contamos cuántas veces terminó en cada estado (frecuencuia de cada estado)
conteos = np.bincount(estados_finales_grande, minlength=8)[1:]  # índice 1-7
frecuencias = conteos / num_simulaciones_grande * 100  # convertir a porcentaje

print("\nDistribución de estados finales tras 100 pasos:")
print(f"{'Estado':<10} {'Veces':<10} {'Frecuencia %':<15}")
print("-" * 35)
for i, (c, f) in enumerate(zip(conteos, frecuencias)):
    barra = "█" * int(f / 2)
    print(f"  {i+1:<7} {c:<10} {f:5.1f}%  {barra}")


TAREA 2: TENDENCIA DE CAMBIO DE LOS ESTADOS (500 simulaciones)

Distribución de estados finales tras 100 pasos:
Estado     Veces      Frecuencia %   
-----------------------------------
  1       208         41.6%  ████████████████████
  2       16           3.2%  █
  3       18           3.6%  █
  4       29           5.8%  ██
  5       15           3.0%  █
  6       26           5.2%  ██
  7       188         37.6%  ██████████████████


Después de 100 iteraciones, un cliente que empieza en estado 4 (neutral) tiende a terminar en:

* Estado 1(Muy insatisfecho): 41.6%
* Estado 7 (Muy satisfecho): 37.6%

El cliente tiene a polarizarse.

## 2.3. Visualización del recorrido

Visualización del recorrido de como cambia el novel de satisfacción, que ilustre como cambia el nivel de satisfacción en cada paso.